File: cs/Parser.cs
namespace Morth;

public static class Parser
{
    public static Op ParseTokenAsOp(Token token)
    {
        if (token.Text == "+")
        {
            return Op.Plus();
        }
        if (token.Text == "-")
        {
            return Op.Minus();
        }
        if (token.Text == ".")
        {
            return Op.Dump();
        }
        if (token.Text == "=")
        {
            return Op.Equal();
        }
        if (token.Text == "if")
        {
            return Op.If();
        }
        if (token.Text == "else")
        {
            return Op.Else();
        }
        if (token.Text == "end")
        {
            return Op.End();
        }
        if (token.Text == "dup")
        {
            return Op.Dup();
        }
        if (token.Text == ">")
        {
            return Op.Greater();
        }
        if (token.Text == "while")
        {
            return Op.While();
        }
        if (token.Text == "do")
        {
            return Op.Do();
        }

        if (!ulong.TryParse(token.Text, out var value))
        {
            throw new ParseException(token, "Unknown token.");
        }

        return Op.Push(value);
    }
}

File: fs/Parser.fs
module Morth.Parser

let parseWord (token : Token.t) =
  match token.word with
  | "+" -> Op.Plus
  | "-" -> Op.Minus
  | "=" -> Op.Eq
  | ">" -> Op.Gt
  | "." -> Op.Dump
  | "if" -> Op.If -1
  | "else" -> Op.Else -1
  | "while" -> Op.While
  | "do" -> Op.Do -1
  | "end" -> Op.End -1
  | "dup" -> Op.Dup
  | n ->
    try
      Op.Push(int n)
    with err ->
      Printf.eprintfn "%s: %s" (token.loc.ToString()) err.Message
      exit 1

let parse (filePath : string) =
  Lexer.lexFile filePath |> Seq.map parseWord |> Blocks.resolve
